codemp/rd-rend2/glsl/ssao.glsl

```diff
@@ -8,95 +8,116 @@ void main()
+
+uniform vec4 u_SSAOSettings; // aocap, 0, aoMultiplier, lightmap
+uniform vec4 u_SSAOSettings2; // 0, aorange, depthTolerance, 0
 
 in vec2 var_ScreenTex;
 
 out vec4 out_Color;
 
-vec2 poissonDisc[9] = vec2[9](
-vec2(-0.7055767, 0.196515),    vec2(0.3524343, -0.7791386),
-vec2(0.2391056, 0.9189604),    vec2(-0.07580382, -0.09224417),
-vec2(0.5784913, -0.002528916), vec2(0.192888, 0.4064181),
-vec2(-0.6335801, -0.5247476),  vec2(-0.5579782, 0.7491854),
-vec2(0.7320465, 0.6317794)
-);
-
-// Input: It uses texture coords as the random number seed.
-// Output: Random number: [0,1), that is between 0.0 and 0.999999... inclusive.
-// Author: Michael Pohoreski
-// Copyright: Copyleft 2012 :-)
-// Source: http://stackoverflow.com/questions/5149544/can-i-generate-a-random-number-inside-a-pixel-shader
-
-float random( const vec2 p )
-{
-  // We need irrationals for pseudo randomness.
-  // Most (all?) known transcendental numbers will (generally) work.
-  const vec2 r = vec2(
-    23.1406926327792690,  // e^pi (Gelfond's constant)
-     2.6651441426902251); // 2^sqrt(2) (Gelfond-Schneider constant)
-  //return fract( cos( mod( 123456789., 1e-7 + 256. * dot(p,r) ) ) );
-  return mod( 123456789., 1e-7 + 256. * dot(p,r) );  
-}
+//
+// AO Shader by Monsterovich :D
+//
 
-mat2 randomRotation( const vec2 p )
-{
-	float r = random(p);
-	float sinr = sin(r);
-	float cosr = cos(r);
-	return mat2(cosr, sinr, -sinr, cosr);
+float readDepth( in vec2 coord, in float znear, in float zfar ) {
+	return (2.0 * znear) / (zfar + znear - texture( u_ScreenDepthMap, coord ).x * (zfar - znear));	
 }
 
-float getLinearDepth(sampler2D depthMap, const vec2 tex, const float zFarDivZNear)
-{
-		float sampleZDivW = texture(depthMap, tex).r;
-		return 1.0 / mix(zFarDivZNear, 1.0, sampleZDivW);
+float compareDepths( in float depth1, in float depth2, in float znear, in float zfar  ) {
+	float diff = sqrt( clamp(1.0-(depth1-depth2) / ( u_SSAOSettings2.y /* aorange */ / (zfar - znear)),0.0,1.0) );
+	float ao = min(u_SSAOSettings.x /* aocap */,max(0.0,depth1-depth2-u_SSAOSettings2.z /* depthTolerance */) * u_SSAOSettings.z /* aoMultiplier */) * diff;
+	return ao;
 }
 
-float ambientOcclusion(sampler2D depthMap, const vec2 tex, const float zFarDivZNear, const float zFar)
+const float P1 = 0.70710678118; // sin,cos(pi/4)
+const float P2x = 0.92387953251; // cos(pi/8)
+const float P2y = 0.38268343236; // sin(pi/8)
+const float P3x = P2y; // cos(3*pi/8)
+const float P3y = P2x; // sin(3*pi/8)
+
+
+void main(void)
 {
-	float result = 0;
-
-	float sampleZ = zFar * getLinearDepth(depthMap, tex, zFarDivZNear);
-
-	vec2 expectedSlope = vec2(dFdx(sampleZ), dFdy(sampleZ)) / vec2(dFdx(tex.x), dFdy(tex.y));
-	
-	if (length(expectedSlope) > 5000.0)
-		return 1.0;
-	
-	vec2 offsetScale = vec2(3.0 / sampleZ);
-	
-	mat2 rmat = randomRotation(tex);
-		
+	float depth = readDepth( var_ScreenTex, u_ViewInfo.x, u_ViewInfo.y );
+
+	float d;
+ 
+	float pw = 1.0 / u_ScreenInfo.x;
```

Comment:
I don't think you need `pw` and `ph`. If you use `texelFetch` to read from the texture, you can use texel coordinates:
```
ivec2 screenCoord = ivec2(g_FragCoord.xy);
vec4 d;
d = texelFetch(u_ScreenDepthMap, screenCoord, 0).x;
d = texelFtech(u_ScreenDepthMap, ivec2(screenCord.x, screenCoord.y + 1, 0).x;
etc..
```

If you really want to continue using `texture`, then `u_ScreenInfo` should contain `(1.0 / w, 1.0 / h)` instead of doing it in the shader.